todo_app/app.py

```diff
@@ -1,54 +1,120 @@
+from todo_app.mongo_db_tasks import TasksDb
+from flask_login import LoginManager, UserMixin, login_required, login_user, logout_user, current_user
 
-app = Flask(__name__)
-headers = {
-   "Accept": "application/json"
-}
+taskDb = TasksDb()
+secret_key = os.environ.get('SECRET_KEY', 'secret_key')
+git_client_id = os.environ.get('GIT_CLIENT_ID')
+git_client_secret = os.environ.get('GIT_CLIENT_SECRET')
+admin_user_id = os.environ.get('APP_ADMIN_USER_ID', '000')
+login_manager = LoginManager()
+client = WebApplicationClient(git_client_id)
+ROLE = 'readwrite'
 
-def create_app():
-    #app = Flask(__name__)
-    #app.config.from_object('dotenv')
+def create_app(taskDb):
+    app = Flask(__name__)    
+    login_disabled = os.environ.get('LOGIN_DISABLED', 'False') == 'True'
+    app.config['LOGIN_DISABLED'] = login_disabled
 
+    #login_manager = LoginManager()
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():        
+        uri = client.prepare_request_uri("https://github.com/login/oauth/authorize")
+        return redirect(uri)
+
+    @login_manager.user_loader    
+    def load_user(user_id):        
+        return User(user_id)
+    
+    @app.route('/login/callback', methods=['GET'])
+    def login_callback():
+        token_response = requests.post('https://github.com/login/oauth/access_token', data={
+                'client_id': git_client_id,
+                'client_secret': git_client_secret,
+                'code': request.args.get('code')
+            }, headers={'Accept': 'application/json'})
+        
+        client.parse_request_body_response(token_response.text)
+
+        user_response = requests.get('https://api.github.com/user', headers={
+            'Accept': 'application/json', 
+            'Authorization': f"token {token_response.json()['access_token']}"
+            })
+        
+        login_user(User(user_response.json()['login']))        
+        return redirect(url_for('index'))
+    
+    def get_user_role():             
+        if (current_app.config.get('LOGIN_DISABLED') or current_user.user_id == admin_user_id):                        
+            return ROLE
+        else:            
+            return 'readonly'
+    
     @app.route('/', methods=['Get'])
+    @login_required
     def index():
-        tasks = trello.get_all_tasks()        
-        task_view_model = view_model.ViewModel(tasks)
+        tasks = taskDb.get_all_tasks()
+        user_role = get_user_role()   
+        task_view_model = view_model.ViewModel(tasks,user_role)
         return render_template('index.html', view_model=task_view_model)
 
+
     @app.route('/add_todo', methods=['Post'])
+    @login_required
     def add_todo_task():
-        trello.create_todo_task(request.form.get('title'))
+        user_role = get_user_role()
+        if (user_role == ROLE):
+            taskDb.create_todo_task(request.form.get('title'))
+            return redirect('/')
+        else:
+            return abort(403)
```

Comment:
Adding this code to each method achieves the goal, but can be a pain to read/maintain. It would be nice to add a custom decorator (e.g. `write_required` or similar, [see this article](https://realpython.com/primer-on-python-decorators/#simple-decorators)) that would run this logic for us